src/Chocolatey.PowerShell/CONTRIBUTING.md

```diff
@@ -0,0 +1,100 @@
+# Contributing Guidelines for `Chocolatey.PowerShell`
+
+This document outlines some guidelines and design practices followed in the `Chocolatey.PowerShell` project.
+It also highlights any important differences from how cmdlets are implemented here compared to the standard patterns for PowerShell cmdlets.
+
+## Naming conventions
+
+Cmdlet classes should be named `VerbNounCommand` and placed in the `Commands` folder and `Chocolatey.PowerShell.Commands` namespace.
+
+## Inherit from `ChocolateyCmdlet` and **not** `PSCmdlet` or `Cmdlet`
+
+`ChocolateyCmdlet` affords some additional helper methods and establishes patterns which can be easily reused across all Chocolatey cmdlets.
+Note that you will still need to apply the standard `[Cmdlet(Verb, Noun)]` attribute on all cmdlet classes for PowerShell to recognise them.
+
+### Overrides
+
+Note that unlike `PSCmdlet`, `ChocolateyCmdlet` requires cmdlets to override the methods `Begin()`, `Process()`, and `End()`. These correspond to `begin {}`, `process {}`, and `end {}` blocks in a PowerShell function.
+
+If unsure, follow these guidelines:
+- Commands that do pipeline processing (declaring a parameter with `[Parameter(ValueFromPipeline = true)]`) will need to handle that input in `Process()`
+- Commands not using pipeline input often handle the bulk of their processing in `End()`
+- If there is any setup that may need to be handled in the command prior to processing any pipeline input, that can go in `Begin()`.
+
+### Logging
+
+Cmdlets inheriting from `ChocolateyCmdlet` will log their parameter values to debug logs when called by default. If any parameters may contain sensitive information, override the `Logging` property and set it to `false` to disable this behaviour.
```

Comment:
Is there any logging that should be done in these cases? Will this be taken care of automatically?

Reply:
At _present_ I would say no. Currently the commands that work this way do not get logged at all I don't think, from the ones I've seen so far?

We could safely turn on the logging after a cmdlet runs (which I think for the commands that have it only mentions the command _name_ and nothing else) and have this property only toggle the logging of parameter values at the start, but I think that would be another divergence in behaviour and we should evaluate that after we finish the rewrite, given I think it has pretty minimal impact whether we log a command after it runs or not.

Most of the commands that do this are... not really part of the public API anyway.